### PR TITLE
Added ContractRegistry

### PR DESCRIPTION
**Pull request by MihanixA**

### CODE REVIEW

File: contracts/utils/VersionControl.sol [hunk 1]
```diff
@@ -0,0 +1,76 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract VersionControl is Ownable {
```
[the remark this answers is not in the document]
Comment (alexeuler) [reply contracts/utils/VersionControl.sol:6]: And I'd make simple: 
EnumerableSet<string> - names
mapping(string => address[]) . The array of deployed contracts.
+ Check how yearn `0x50c1a2eA0a861A967D9d0FFE2AE4012c2E053804` registry works, probably smth useful there

[the remark this answers is not in the document]
Comment (alexeuler) [reply contracts/utils/VersionControl.sol:6]: is Ownable => is DefaultAccessControl

File: contracts/vaults/AaveVault.sol [hunk 2]
```diff
@@ -25,8 +25,10 @@ import "./IntegrationVault.sol";
 /// It is assumed that any amounts of tokens can be deposited / withdrawn from Aave.
 /// The contract's vaultTokens are fully allowed to Aave Lending Pool.
 contract AaveVault is IAaveVault, IntegrationVault {
-    address[] internal _aTokens;
-    uint256[] internal _tvls;
+    string public constant VERSION = "1.0.0";
```
Comment (alexeuler) [inline contracts/vaults/AaveVault.sol:28]: I though we discussed to update only Governances code?

File: contracts/vaults/AggregateVault.sol [hunk 3]
```diff
@@ -12,7 +12,7 @@ import "../libraries/ExceptionsLibrary.sol";
 import "../libraries/PermissionIdsLibrary.sol";
 
 /// @notice Vault that combines several integration layer Vaults into one Vault.
-contract AggregateVault is IAggregateVault, Vault {
+abstract contract AggregateVault is IAggregateVault, Vault {
```
Comment (alexeuler) [inline contracts/vaults/AggregateVault.sol:15]: I think this is unnecessary - it'll be easier to test without abstract

File: contracts/vaults/YearnVaultGovernance.sol [hunk 4]
```diff
@@ -8,6 +8,8 @@ import "./VaultGovernance.sol";
 
 /// @notice Governance that manages all Aave Vaults params and can deploy a new Aave Vault.
 contract YearnVaultGovernance is IYearnVaultGovernance, VaultGovernance {
+    string public constant VERSION = "1.0.0";
```
Comment (alexeuler) [inline contracts/vaults/YearnVaultGovernance.sol:11]: Also let's add for oracles, strategies and other meaningful contracts

**Review by alexeuler — CHANGES_REQUESTED** [review]

File: contracts/vaults/AaveVault.sol (hunk 2 above)
Comment (MihanixA) [reply contracts/vaults/AaveVault.sol]: I thought we decided to add it vaults as well at the end

**Comment by MihanixA** [comment]
[Add version to contracts and feature to query latest version](https://trello.com/c/fGtPnZtY/286-add-version-to-contracts-and-feature-to-query-latest-version)

File: contracts/libraries/SemverLibrary.sol [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    function numberify(bytes memory _num) internal pure returns (uint256 result) {
+        for (uint256 i; i != _num.length; ++i) {
+            result *= 10;
+            result += uint256(uint8(_num[i])) - ASCII_ZERO;
+        }
+    }
+
+    function stringify(uint256 num) internal pure returns (bytes memory result) {
```
Comment (alexeuler) [inline contracts/libraries/SemverLibrary.sol:31]: There's Strings.toString in openzeppelin utils

File: contracts/libraries/SemverLibrary.sol [hunk 6]
```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+library SemverLibrary {
```
Comment (alexeuler) [inline contracts/libraries/SemverLibrary.sol:4]: Can we make this simpler by assuming that semver is always x.y.z where numbers are 0 to 255?

File: contracts/libraries/SemverLibrary.sol [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            num /= 10;
+            ++i;
+        }
+        result = new bytes(i);
+        for (uint256 j; j != i; ++j) {
+            result[j] = resultTemp[i - j - 1];
+        }
+    }
+
+    function numberifySemver(bytes32 _semver) internal pure returns (uint256) {
```
Comment (alexeuler) [inline contracts/libraries/SemverLibrary.sol:49]: While StateMachine is a good pattern in some cases it's not very clear for verification.
What I'd do here is find dot indexes and then pass that to numberify method. Looks like that way it'll be simpler

File: contracts/utils/DefaultAccessControl.sol [hunk 8]
```diff
@@ -8,17 +8,21 @@ import "../libraries/ExceptionsLibrary.sol";
...
         require(admin != address(0), ExceptionsLibrary.ADDRESS_ZERO);
-        bytes32 adminRole = ADMIN_ROLE;
-        _setupRole(adminRole, admin);
-        _setRoleAdmin(adminRole, adminRole);
-        _setRoleAdmin(ADMIN_DELEGATE_ROLE, adminRole);
+
+        _setupRole(OPERATOR, admin);
+        _setupRole(ADMIN_ROLE, admin);
+
+        _setRoleAdmin(OPERATOR, ADMIN_ROLE);
```
Comment (alexeuler) [inline contracts/utils/DefaultAccessControl.sol:23]: ADMIN_DELEGATE should rule operators. ADMIN is a sort of SUPER_ADMIN

**Review by alexeuler — CHANGES_REQUESTED** [review]

File: contracts/ContractRegistry.sol [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            ExceptionsLibrary.FORBIDDEN
+        );
+        for (uint256 i; i != targets.length; ++i) {
+            require(_registeredContracts.add(targets[i]), ExceptionsLibrary.DUPLICATE);
+            IContractMeta newContract = IContractMeta(targets[i]);
+            bytes32 newContractName = newContract.CONTRACT_NAME();
+            uint256 newContractVersion = newContract.CONTRACT_VERSION();
+            registeredContractsAddresses[newContractName].push(targets[i]);
+            uint256 currentVersionCount = versionCount[newContractName];
+            require(currentVersionCount + 1 >= newContractVersion, ExceptionsLibrary.INVALID_VALUE);
```
[the remark this answers is not in the document]
Comment (MihanixA) [reply contracts/ContractRegistry.sol]: outdated.

File: contracts/libraries/SemverLibrary.sol (hunk 6 above)
Comment (MihanixA) [reply contracts/libraries/SemverLibrary.sol:4]: no

File: contracts/libraries/SemverLibrary.sol (hunk 5 above)
Comment (MihanixA) [reply contracts/libraries/SemverLibrary.sol:31]: strings are immutable 
I don't want to do an extra cast

**Review by alexeuler — APPROVED** [review]

**Pull request merged**